### PR TITLE
Add a Nova workflow to run iOS test on AWS Device Farm

### PR DESCRIPTION
**Pull request by huydhn**
This adds a new Nova workflow called `mobile_job` that can be used to run app tests on AWS Device Farm

1. In the initial version, the workflow will support running xctest suite for iOS app.  It requires the iOS IPA app archive and xctest suite as inputs per https://docs.aws.amazon.com/devicefarm/latest/developerguide/test-types-ios-xctest.html.  For example,
```
jobs:
  test-mobile-job:
    permissions:
      id-token: write
      contents: read
    uses: ./.github/workflows/mobile_job.yml
    with:
      device-type: ios
      runner: ubuntu-latest
      # This is PyTorch project on Device Farm
      project-arn: arn:aws:devicefarm:us-west-2:308535385114:project:b531574a-fb82-40ae-b687-8f0b81341ae0
      # The caller workflow will need to provide these artifacts. GHA support could be added later together with S3
      ios-ipa-archive: https://ossci-assets.s3.amazonaws.com/TestApp.ipa
      ios-xctest-zip: https://ossci-assets.s3.amazonaws.com/TestAppTests.xctest.zip
```

2. The main logic is in the script `tools/device-farm-runner/run_on_aws_devicefarm.py` where it uploads the IPA app archive and xctest suite to Device Farm, pools for the results, and reports back.  For example,

```
python run_on_aws_devicefarm.py --project-arn arn:aws:devicefarm:us-west-2:308535385114:project:b531574a-fb82-40ae-b687-8f0b81341ae0  --app-file ~/TestApp.ipa --xctest-file ~/TestAppTests.xctest.zip --name-prefix debug
```

3. The permission to Device Farm is granted by the role `gha_workflow_mobile_job` from https://github.com/pytorch-labs/pytorch-gha-infra/pull/348

### Testing

On CI, I setup a test workflow to run PyTorch iOS app tests https://github.com/pytorch/test-infra/actions/runs/8074168433/job/22059038480?pr=4973

Locally,

* If the run on Device Farm passes:
```
Run debug-invalid-workflow-id-0-2024-02-27-qVMFnLAf is scheduled as arn:aws:devicefarm:us-west-2:308535385114:run:b531574a-fb82-40ae-b687-8f0b81341ae0/295735c2-2a07-46f8-be67-ef9604a62d17
mobile-job-ios-8060493871-2-2024-02-27-IdKPJMzZ PASSED with stats {'total': 195, 'passed': 195, 'failed': 0, 'warned': 0, 'errored': 0, 'stopped': 0, 'skipped': 0}
```

* If the run on Device Farm fails, the failed tests will be print together with the link to their artifacts on S3 for manual debugging:
```
Run debug-invalid-workflow-id-0-2024-02-27-XNivrOMS is scheduled as arn:aws:devicefarm:us-west-2:308535385114:run:b531574a-fb82-40ae-b687-8f0b81341ae0/fea59a70-c1ef-41ac-81c3-b8225b4aecf1
PyTorch-2023-09-29-OeFtuHCz FAILED
  Apple iPad Pro 12.9" (2022) PASSED
  Apple iPhone 11 PASSED
  Apple iPhone 13 FAILED
    Setup Suite PASSED with stats {'total': 1, 'passed': 1, 'failed': 0, 'warned': 0, 'errored': 0, 'stopped': 0, 'skipped': 0}
    TestAppTests FAILED with stats {'total': 37, 'passed': 36, 'failed': 1, 'warned': 0, 'errored': 0, 'stopped': 0, 'skipped': 0}
      testTorchScriptCollectionQuantOps PASSED
      testTorchScriptBuiltinQuantOps PASSED
      testFusedQuantOps PASSED
      testStaticQuantOps PASSED
      testDynamicQuantOps PASSED
      testQuantOps PASSED
      testNNUtilsOps PASSED
      testShuffleOps PASSED
      testVisionFunctionOps PASSED
      testLossFunctionOps PASSED
      testDistanceFunctionOps PASSED
      testSparseOps PASSED
      testDropoutOps PASSED
      testLinearOps PASSED
      testTransformerOps PASSED
      testRecurrentOps PASSED
      testNormalizationOps PASSED
      testActivationOps PASSED
      testPaddingOps PASSED
      testPoolingOps PASSED
      testConvolutionOps PASSED
      testTensorViewOps PASSED
      testTensorTypingOps PASSED
      testTensorIndexingOps PASSED
      testTensorCreationOps PASSED
      testTensorOps PASSED
      testSamplingOps PASSED
      testBlasLapackOps PASSED
      testSpectralOps PASSED
      testOtherMathOps PASSED
      testComparisonOps PASSED
      testReductionOps PASSED
      testPointwiseOps PASSED
      testMobileNetV2 PASSED
      testLiteInterpreter PASSED
      testModel PASSED
      testCoreML FAILED
        Saving FILE Automation_Output.txt (AUTOMATION_OUTPUT) at https://gha-artifacts.s3.amazonaws.com/device_farm/invalid-workflow-id/0/arn_aws_devicefarm_us-west-2_308535385114_artifact_b531574a-fb82-40ae-b687-8f0b81341ae0_fea59a70-c1ef-41ac-81c3-b8225b4aecf1_00003_00001_00036_00000_Automation_Output.txt
        Saving FILE Application_Output.txt (UNKNOWN) at https://gha-artifacts.s3.amazonaws.com/device_farm/invalid-workflow-id/0/arn_aws_devicefarm_us-west-2_308535385114_artifact_b531574a-fb82-40ae-b687-8f0b81341ae0_fea59a70-c1ef-41ac-81c3-b8225b4aecf1_00003_00001_00036_00001_Application_Output.txt
        Saving FILE Application_Crash_Report.ips (APPLICATION_CRASH_REPORT) at https://gha-artifacts.s3.amazonaws.com/device_farm/invalid-workflow-id/0/arn_aws_devicefarm_us-west-2_308535385114_artifact_b531574a-fb82-40ae-b687-8f0b81341ae0_fea59a70-c1ef-41ac-81c3-b8225b4aecf1_00003_00001_00036_00002_Application_Crash_Report.ips
        Saving FILE Video.mp4 (VIDEO) at https://gha-artifacts.s3.amazonaws.com/device_farm/invalid-workflow-id/0/arn_aws_devicefarm_us-west-2_308535385114_artifact_b531574a-fb82-40ae-b687-8f0b81341ae0_fea59a70-c1ef-41ac-81c3-b8225b4aecf1_00003_00001_00036_00003_Video.mp4
        Saving FILE Syslog.syslog (DEVICE_LOG) at https://gha-artifacts.s3.amazonaws.com/device_farm/invalid-workflow-id/0/arn_aws_devicefarm_us-west-2_308535385114_artifact_b531574a-fb82-40ae-b687-8f0b81341ae0_fea59a70-c1ef-41ac-81c3-b8225b4aecf1_00003_00001_00036_00004_Syslog.syslog
        Saving FILE TCP_dump_log.txt (RAW_FILE) at https://gha-artifacts.s3.amazonaws.com/device_farm/invalid-workflow-id/0/arn_aws_devicefarm_us-west-2_308535385114_artifact_b531574a-fb82-40ae-b687-8f0b81341ae0_fea59a70-c1ef-41ac-81c3-b8225b4aecf1_00003_00001_00036_00005_TCP_dump_log.txt
        Saving LOG ListArtifactType.log.json (MESSAGE_LOG) at https://gha-artifacts.s3.amazonaws.com/device_farm/invalid-workflow-id/0/arn_aws_devicefarm_us-west-2_308535385114_artifact_b531574a-fb82-40ae-b687-8f0b81341ae0_fea59a70-c1ef-41ac-81c3-b8225b4aecf1_00003_00001_00036_LOG_ListArtifactType.log.json
    Teardown Suite PASSED with stats {'total': 1, 'passed': 1, 'failed': 0, 'warned': 0, 'errored': 0, 'stopped': 0, 'skipped': 0}
  Apple iPhone 14 Pro PASSED
  Apple iPhone SE (2022) PASSED
```

### CODE REVIEW

File: .github/workflows/mobile_job.yml [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+          echo "::group::Cleanup debug output"
+          rm -rfv "${GITHUB_WORKSPACE}"
+          mkdir -p "${GITHUB_WORKSPACE}"
+          echo "::endgroup::"
+
+      - name: Authenticate with AWS
+        uses: aws-actions/configure-aws-credentials@v3
+        with:
+          # TODO: Replace with an actual role from pytorch-gha-infra
+          role-to-assume: arn:aws:iam::308535385114:role/DEBUG-DEVICE-FARM-ROLE
```
Comment (huydhn) [inline .github/workflows/mobile_job.yml:76]: This is to be replaced by TODO

**Comment by atalman** [comment]
This is nice! Can you provide some context/use case where are we planning on using this workflow ?

**Comment by huydhn** [comment]
> This is nice! Can you provide some context/use case where are we planning on using this workflow ?

This would initially be used by ExecuTorch to run their iOS tests.  Atm there are only some correctness tests in their demo app, however, the next step would be to work with @shoumikhin to run some kind of benchmark on devices.  IMO, it could start off small with a simple inference for loop to measure runtime on devices like what PyTorch mobile has https://github.com/pytorch/pytorch/blob/main/ios/TestApp/TestApp/Benchmark.mm#L78-L97.

As an example, I will onboard PyTorch iOS workflow to use this.  The test workflow you see in the PR actually uses PyTorch iOS demo app.

File: .github/workflows/mobile_job.yml (hunk 1 above)
Comment (seemethere) [reply .github/workflows/mobile_job.yml:76]: We will also probably want to make a follow up issue to ensure this doesn't break if / when we migrate to ARC

Comment (jeanschmidt) [reply .github/workflows/mobile_job.yml]: The credentials are acquired at user/workflow level. Given that ARC provides the same version of AWS cli/toolset I believe that there are no changes required. Unless there are permissions granted to the node role that won't be available in ARC. But I've been disencouraging it for a while, so I believe we don't have such cases ATM

**Review by seemethere — APPROVED** [review]

File: tools/device-farm-runner/run_on_aws_devicefarm.py [hunk 2]
```diff
@@ -0,0 +1,336 @@
...
+
+    with open(filename, "rb") as file_stream:
+        info(f"Uploading {filename} to Device Farm as {upload_name}...")
+        r = requests.put(upload_url, data=file_stream, headers={"content-type": mime})
+        if not r.ok:
+            raise Exception(f"Couldn't upload {filename}: {r.reason}")
+
+    start_time = datetime.datetime.now()
+    # Polling AWS till the uploaded file is ready
+    while True:
```
Comment (malfet) [inline tools/device-farm-runner/run_on_aws_devicefarm.py:115]: `while True` is kind of movetón :)
```suggestion
    while status != "SUCCESS":
```

File: tools/device-farm-runner/run_on_aws_devicefarm.py [hunk 3]
```diff
@@ -0,0 +1,336 @@
...
+        if status == "SUCCEEDED":
+            break
+
+        time.sleep(POLLING_DELAY_IN_SECOND)
+
+    return upload_arn
+
+
+def is_success(result: Optional[str]) -> bool:
+    return bool(result and result == "PASSED")
```
Comment (malfet) [inline tools/device-farm-runner/run_on_aws_devicefarm.py:138]: Nit (None is not equal to any string)
```suggestion
    return result == "PASSED"
```

**Pull request merged**